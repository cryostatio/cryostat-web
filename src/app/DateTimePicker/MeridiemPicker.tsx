/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { css } from '@patternfly/react-styles';
import * as React from 'react';
import { useTranslation } from 'react-i18next';

export interface MeridiemPickerProps {
  onSelect?: (isAM: boolean) => void;
  isAM?: boolean;
}

export const MeridiemPicker: React.FC<MeridiemPickerProps> = ({ onSelect = () => undefined, isAM = true }) => {
  const { t } = useTranslation();
  const handleSelectAM = React.useCallback(() => {
    onSelect(true);
  }, [onSelect]);

  const handleSelectPM = React.useCallback(() => {
    onSelect(false);
  }, [onSelect]);

  return (
    <div
      role={'listbox'}
      aria-label={t('MeridiemPicker.ARIA_LABELS.LISTBOX') || ''}
      className="datetime-picker__meridiem-title-stack"
    >
      <div className={css('datetime-picker__meridiem-tile', `${isAM ? '' : 'un'}selected`)} onClick={handleSelectAM}>
        {t('MERIDIEM_AM', { ns: 'common' })}
      </div>
      <div className={css('datetime-picker__meridiem-tile', `${isAM ? 'un' : ''}selected`)} onClick={handleSelectPM}>
        {t('MERIDIEM_PM', { ns: 'common' })}
      </div>
    </div>
  );
};
