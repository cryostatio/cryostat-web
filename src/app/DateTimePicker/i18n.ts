/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * t('TimeSpinner.INPUT_HOUR12_VALUE')
 * t('TimeSpinner.INPUT_HOUR24_VALUE')
 * t('TimeSpinner.INPUT_MINUTE_VALUE')
 * t('TimeSpinner.INPUT_SECOND_VALUE')
 * t('TimeSpinner.INCREMENT_HOUR12_VALUE')
 * t('TimeSpinner.INCREMENT_HOUR24_VALUE')
 * t('TimeSpinner.INCREMENT_MINUTE_VALUE')
 * t('TimeSpinner.INCREMENT_SECOND_VALUE')
 * t('TimeSpinner.DECREMENT_HOUR12_VALUE')
 * t('TimeSpinner.DECREMENT_HOUR24_VALUE')
 * t('TimeSpinner.DECREMENT_MINUTE_VALUE')
 * t('TimeSpinner.DECREMENT_SECOND_VALUE')
 */
