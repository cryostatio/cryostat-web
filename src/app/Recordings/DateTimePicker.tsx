/*
 * Copyright The Cryostat Authors
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 *
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */


import {
  Button,
  ButtonVariant,
  DatePicker,
  Flex,
  FlexItem,
  InputGroup,
  isValidDate,
  Text,
  TimePicker,
  yyyyMMddFormat,
} from '@patternfly/react-core';
import { SearchIcon } from '@patternfly/react-icons';
import React from 'react';

export interface DateTimePickerProps {
  onSubmit: (startDate) => void;
}

export const DateTimePicker: React.FunctionComponent<DateTimePickerProps> = (props) => {
  const [start, setStart] = React.useState(new Date(0));
  const [searchDisabled, setSearchDisabled] = React.useState(true);

  const onStartDateChange = React.useCallback(
    (inputDate, newStartDate) => {
      if (isValidDate(start) && isValidDate(newStartDate) && inputDate === yyyyMMddFormat(newStartDate)) {
        setStart(new Date(newStartDate));
        setSearchDisabled(false);
      } else {
        setSearchDisabled(true);
      }
    },
    [start, isValidDate, yyyyMMddFormat]
  );

  const onStartTimeChange = React.useCallback(
    (unused, hour, minute) => {
      let updated = new Date(start);
      updated.setUTCHours(hour, minute);
      setStart(updated);
    },
    [start, setStart, isValidDate]
  );

  const handleSubmit = React.useCallback(() => {
    props.onSubmit(`${start.toISOString()}`);
  }, [start, props.onSubmit]);

  return (
    <Flex>
      <FlexItem>
        <InputGroup>
          <DatePicker
            onChange={onStartDateChange}
            aria-label="Start date"
            placeholder="YYYY-MM-DD" />
          <TimePicker is24Hour aria-label="Start time" className="time-picker" onChange={onStartTimeChange} />
        </InputGroup>
        </FlexItem>
        <FlexItem>
          <Text>UTC</Text>
        </FlexItem>
        <FlexItem>
          <Button variant={ButtonVariant.control} aria-label="search button for start date" isDisabled={searchDisabled} onClick={handleSubmit}>
            <SearchIcon />
          </Button>
      </FlexItem>
    </Flex>
  );
};
