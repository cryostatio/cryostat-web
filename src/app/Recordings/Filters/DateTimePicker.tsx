/*
 * Copyright The Cryostat Authors
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 *
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import {
  Button,
  ButtonVariant,
  DatePicker,
  Flex,
  FlexItem,
  InputGroup,
  Text,
  TimePicker,
} from '@patternfly/react-core';
import { SearchIcon } from '@patternfly/react-icons';
import React from 'react';

export interface DateTimePickerProps {
  onSubmit: (date: any) => void;
}

export const DateTimePicker: React.FunctionComponent<DateTimePickerProps> = (props) => {
  const [selectedDate, setSelectedDate] = React.useState<Date | undefined>();
  const [selectedHour, setSelectedHour] = React.useState(0);
  const [selectedMinute, setSelectedMinute] = React.useState(0);
  const [isTimeOpen, setIsTimeOpen] = React.useState(false);

  const onDateChange = React.useCallback(
    (inputDateValue: string, newDate: Date | undefined) => {
      setSelectedDate(newDate);
    },
    [setSelectedDate]
  );

  const onTimeChange = React.useCallback(
    (_, hour, minute) => {
      setSelectedHour(hour);
      setSelectedMinute(minute);
    },
    [setSelectedHour, setSelectedMinute]
  );

  const onTimeToggle = React.useCallback(
    (opened) => {
      setIsTimeOpen(opened);
    },
    [setIsTimeOpen]
  );

  const handleSubmit = React.useCallback(() => {
    selectedDate!.setUTCHours(selectedHour, selectedMinute);
    props.onSubmit(selectedDate!.toISOString());
  }, [selectedDate, selectedHour, selectedMinute, props.onSubmit]);

  return (
    <Flex>
      <FlexItem>
        <InputGroup>
          <DatePicker appendTo="parent" onChange={onDateChange} aria-label="Date Picker" placeholder="YYYY-MM-DD" />
          <TimePicker
            isOpen={isTimeOpen}
            setIsOpen={onTimeToggle}
            is24Hour
            aria-label="Time Picker"
            className="time-picker"
            menuAppendTo="parent"
            onChange={onTimeChange}
          />
        </InputGroup>
      </FlexItem>
      <FlexItem>
        <Text>UTC</Text>
      </FlexItem>
      <FlexItem>
        <Button
          variant={ButtonVariant.control}
          aria-label="Search For Date"
          isDisabled={!selectedDate}
          onClick={handleSubmit}
        >
          <SearchIcon />
        </Button>
      </FlexItem>
    </Flex>
  );
};
