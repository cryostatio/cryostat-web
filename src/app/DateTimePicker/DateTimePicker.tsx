/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { TimePicker } from '@app/DateTimePicker/TimePicker';
import { useDayjs } from '@app/utils/hooks/useDayjs';
import { Timezone, defaultDatetimeFormat } from '@i18n/datetime';
import { isHourIn24hAM } from '@i18n/datetimeUtils';
import {
  ActionGroup,
  Bullseye,
  Button,
  CalendarMonth,
  Flex,
  FlexItem,
  Form,
  FormGroup,
  Tab,
  Tabs,
  TabTitleText,
  TextInput,
} from '@patternfly/react-core';
import * as React from 'react';
import { useTranslation } from 'react-i18next';
import { TimezonePicker } from './TimezonePicker';

export interface DateTimePickerProps {
  onSelect?: (date: Date, timezone: Timezone) => void;
  prefilledDate?: Date; // timezone is ignored
  onDismiss: () => void;
}

type _TabKey = 'date' | 'time';

export const DateTimePicker: React.FC<DateTimePickerProps> = ({ onSelect, onDismiss, prefilledDate }) => {
  const { t } = useTranslation();
  const [dayjs, _] = useDayjs();
  const [activeTab, setActiveTab] = React.useState<_TabKey>('date');
  const [datetime, setDatetime] = React.useState<Date>(new Date());
  const [timezone, setTimezone] = React.useState<Timezone>(defaultDatetimeFormat.timeZone); // Not affected by user preferences

  const handleTabSelect = React.useCallback(
    (_: MouseEvent | React.MouseEvent, key: string | number) => setActiveTab(`${key}` as _TabKey),
    [setActiveTab],
  );

  const handleSubmit = React.useCallback(() => {
    onSelect && onSelect(datetime, timezone);
  }, [datetime, timezone, onSelect]);

  const handleCalendarSelect = React.useCallback(
    (_, date: Date) => {
      setDatetime((old) => {
        const wrappedOld = dayjs(old);
        return dayjs(date).hour(wrappedOld.hour()).minute(wrappedOld.minute()).second(wrappedOld.second()).toDate();
      });
      setActiveTab('time'); // Switch to time
    },
    [setDatetime, setActiveTab, dayjs],
  );

  const handleHourChange = React.useCallback(
    (hrIn24h: number) => {
      setDatetime((old) => dayjs(old).hour(hrIn24h).toDate());
    },
    [setDatetime, dayjs],
  );

  const handleMinuteChange = React.useCallback(
    (m: number) => {
      setDatetime((old) => dayjs(old).minute(m).toDate());
    },
    [setDatetime, dayjs],
  );

  const handleSecondChange = React.useCallback(
    (s: number) => {
      setDatetime((old) => dayjs(old).second(s).toDate());
    },
    [setDatetime, dayjs],
  );

  const handleMeridiemChange = React.useCallback(
    (isAM: boolean) => {
      setDatetime((old) => {
        const oldDayjs = dayjs(old);
        if (isAM !== isHourIn24hAM(oldDayjs.hour())) {
          return (isAM ? oldDayjs.subtract(12, 'hour') : oldDayjs.add(12, 'hour')).toDate();
        }
        return old;
      });
    },
    [setDatetime, dayjs],
  );

  const selectedDatetimeDisplay = React.useMemo(() => dayjs(datetime).format('L LTS'), [datetime, dayjs]);

  React.useEffect(() => {
    if (prefilledDate) {
      setDatetime(new Date(prefilledDate));
    }
  }, [setDatetime, prefilledDate]);

  return (
    <Form>
      <Tabs
        aria-label={t('DateTimePicker.ARIA_LABELS.TABS') || ''}
        onSelect={handleTabSelect}
        activeKey={activeTab}
        isFilled
        role="region"
      >
        <Tab key={'date'} eventKey={'date'} title={<TabTitleText>{t('DATE', { ns: 'common' })}</TabTitleText>}>
          <FormGroup>
            <Bullseye>
              <CalendarMonth
                className="datetime-picker__calendar"
                isDateFocused
                locale={dayjs.locale()}
                inlineProps={{
                  component: 'article',
                  ariaLabelledby: 'start-date',
                }}
                style={{ padding: 0 }}
                date={datetime}
                onChange={handleCalendarSelect}
              />
            </Bullseye>
          </FormGroup>
        </Tab>
        <Tab key={'time'} eventKey={'time'} title={<TabTitleText>{t('TIME', { ns: 'common' })}</TabTitleText>}>
          <FormGroup>
            <Bullseye>
              <TimePicker
                selected={{
                  hour24: dayjs(datetime).hour(), // 24hr format
                  minute: dayjs(datetime).minute(),
                  second: dayjs(datetime).second(),
                }}
                onHourSelect={handleHourChange}
                onMinuteSelect={handleMinuteChange}
                onSecondSelect={handleSecondChange}
                onMeridiemSelect={handleMeridiemChange}
              />
            </Bullseye>
          </FormGroup>
        </Tab>
      </Tabs>
      <FormGroup label={t('DateTimePicker.SELECTED_DATETIME')}>
        <Flex>
          <FlexItem>
            <TextInput
              id="selected-datetime"
              aria-label={t('DateTimePicker.ARIA_LABELS.DISPLAY_SELECTED_DATETIME') || ''}
              className="datetime-picker__datetime-selected-display"
              readOnly
              value={selectedDatetimeDisplay}
            />
          </FlexItem>
          <FlexItem>
            <TimezonePicker menuAppendTo={document.body} onTimezoneChange={setTimezone} selected={timezone} isCompact />
          </FlexItem>
        </Flex>
      </FormGroup>
      <ActionGroup style={{ marginTop: 0 }}>
        <Button variant="primary" onClick={handleSubmit}>
          {t('SELECT', { ns: 'common' })}
        </Button>
        <Button variant="link" onClick={onDismiss}>
          {t('CANCEL', { ns: 'common' })}
        </Button>
      </ActionGroup>
    </Form>
  );
};
