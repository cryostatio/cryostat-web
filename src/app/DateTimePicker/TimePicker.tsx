/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { MeridiemPicker } from '@app/DateTimePicker/MeridiemPicker';
import { format2Digit, hourIn12HrFormat, hourIn24HrFormat, isHourIn24hAM } from '@i18n/datetimeUtils';
import {
  Button,
  Divider,
  HelperText,
  HelperTextItem,
  Icon,
  Level,
  LevelItem,
  Panel,
  PanelFooter,
  PanelMain,
  PanelMainBody,
  Stack,
  StackItem,
  Switch,
  TextInput,
  Title,
} from '@patternfly/react-core';
import { AngleDownIcon, AngleUpIcon } from '@patternfly/react-icons';
import { css } from '@patternfly/react-styles';
import _ from 'lodash';
import * as React from 'react';
import { useTranslation } from 'react-i18next';

export interface TimePickerProps {
  selected: {
    // controlled
    hour24: number; // In 24h format
    minute: number;
    second: number;
  };
  onHourSelect?: (hrIn24h: number) => void;
  onMinuteSelect?: (minute: number) => void;
  onSecondSelect?: (second: number) => void;
  onMeridiemSelect?: (isAM: boolean) => void;
}

export const TimePicker: React.FC<TimePickerProps> = ({
  onHourSelect,
  onMinuteSelect,
  onSecondSelect,
  onMeridiemSelect,
  selected,
}) => {
  const { t } = useTranslation();
  const [is24h, setIs24h] = React.useState(true);

  const meridiemAM = React.useMemo(() => isHourIn24hAM(selected.hour24), [selected.hour24]);

  const handleHourSelect = React.useCallback(
    (rawHour: number) => {
      const hour = is24h ? rawHour : hourIn24HrFormat(rawHour, meridiemAM);
      onHourSelect && onHourSelect(hour);
    },
    [is24h, onHourSelect, meridiemAM],
  );

  return (
    <>
      <Panel>
        <PanelMain>
          <PanelMainBody>
            <Level hasGutter>
              <LevelItem key={'hour'}>
                <TimeSpinner
                  variant={is24h ? 'hour24' : 'hour12'}
                  label={t('HOUR', { ns: 'common' })}
                  selected={is24h ? selected.hour24 : hourIn12HrFormat(selected.hour24)[0]}
                  onChange={handleHourSelect}
                />
              </LevelItem>
              <LevelItem key={'splitter-1'}>
                <div className="datetime-picker__colon-divider">:</div>
              </LevelItem>
              <LevelItem key={'minute'}>
                <TimeSpinner
                  variant={'minute'}
                  label={t('MINUTE', { ns: 'common' })}
                  selected={selected.minute}
                  onChange={onMinuteSelect}
                />
              </LevelItem>
              <LevelItem key={'splitter-2'}>
                <div className="datetime-picker__colon-divider">:</div>
              </LevelItem>
              <LevelItem key={'second'}>
                <TimeSpinner
                  variant={'second'}
                  label={t('SECOND', { ns: 'common' })}
                  selected={selected.second}
                  onChange={onSecondSelect}
                />
              </LevelItem>
              {is24h ? (
                <></>
              ) : (
                <LevelItem key={'meridiem'}>
                  <MeridiemPicker isAM={meridiemAM} onSelect={onMeridiemSelect} />
                </LevelItem>
              )}
            </Level>
          </PanelMainBody>
          <Divider />
          <PanelFooter>
            <HelperText>
              <HelperTextItem>{t('TimePicker.USE_24HR_TIME')}</HelperTextItem>
            </HelperText>
            <Switch
              id={'24-hour-switch'}
              label={t('TimePicker.24HOUR')}
              isChecked={is24h}
              onChange={(_, checked: boolean) => setIs24h(checked)}
            />
          </PanelFooter>
        </PanelMain>
      </Panel>
    </>
  );
};

interface TimeSpinnerProps {
  variant: 'hour12' | 'hour24' | 'minute' | 'second';
  onChange?: (value: number) => void;
  selected: number; // controlled, must be corresponding to variant
  label?: string | null;
}

const TimeSpinner: React.FC<TimeSpinnerProps> = ({ variant, onChange, selected, label }) => {
  const { t } = useTranslation();
  const computedMax = React.useMemo(() => {
    switch (variant) {
      case 'hour12':
        return 12;
      case 'hour24':
        return 23;
      default: // minute, second
        return 59;
    }
  }, [variant]);

  const computedMin = React.useMemo(() => {
    switch (variant) {
      case 'hour12':
        return 1;
      default:
        return 0; // hour24, minute, second
    }
  }, [variant]);

  const _sanitizeValue = React.useCallback(
    (value: number) => {
      return _.clamp(value, computedMin, computedMax);
    },
    [computedMax, computedMin],
  );

  const handleValueChange = React.useCallback(
    (_, value: string) => {
      if (isNaN(Number(value))) {
        return;
      }
      const newVal = _sanitizeValue(Number(value));
      onChange && onChange(newVal);
    },
    [onChange, _sanitizeValue],
  );

  const handleIncrement = React.useCallback(() => {
    const newVal = _sanitizeValue(selected + 1);
    onChange && onChange(newVal);
  }, [selected, _sanitizeValue, onChange]);

  const handleDecrement = React.useCallback(() => {
    const newVal = _sanitizeValue(selected - 1);
    onChange && onChange(newVal);
  }, [selected, _sanitizeValue, onChange]);

  return (
    <Stack aria-label={label || variant}>
      {label ? (
        <StackItem>
          <Title className="datetime-picker__time-text-top-label" headingLevel={'h4'}>
            {label}
          </Title>
        </StackItem>
      ) : (
        <></>
      )}
      <StackItem key={`${variant}-increment`}>
        <Button
          className={css('datetime-picker__time-spin-box', 'up')}
          onClick={handleIncrement}
          aria-label={t(`TimeSpinner.INCREMENT_${variant.toUpperCase()}_VALUE`) || ''}
        >
          <Icon size="md">
            <AngleUpIcon />
          </Icon>
        </Button>
      </StackItem>
      <StackItem key={`${variant}-input`}>
        <TextInput
          id={`${variant}-input`}
          aria-label={t(`TimeSpinner.INPUT_${variant.toUpperCase()}_VALUE`) || ''}
          className="datetime-picker__number-input"
          type="number"
          min={computedMin}
          max={computedMax}
          value={format2Digit(selected)}
          onChange={handleValueChange}
        />
      </StackItem>
      <StackItem key={`${variant}-decrement`}>
        <Button
          className={css('datetime-picker__time-spin-box', 'down')}
          onClick={handleDecrement}
          aria-label={t(`TimeSpinner.DECREMENT_${variant.toUpperCase()}_VALUE`) || ''}
        >
          <Icon size="md">
            <AngleDownIcon />
          </Icon>
        </Button>
      </StackItem>
    </Stack>
  );
};
