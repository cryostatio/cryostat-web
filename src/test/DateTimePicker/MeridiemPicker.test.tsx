/*
 * Copyright The Cryostat Authors
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 *
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { MeridiemPicker } from '@app/DateTimePicker/MeridiemPicker';
import { cleanup, screen } from '@testing-library/react';
import * as React from 'react';
import renderer, { act } from 'react-test-renderer';
import { renderDefault, testTranslate } from '../Common';

const onSelect = jest.fn((_: boolean) => undefined);

describe('<MeridiemPicker/>', () => {
  beforeEach(() => {
    jest.mocked(onSelect).mockReset();
  });

  afterEach(cleanup);

  it('renders correctly', async () => {
    let tree;
    await act(async () => {
      tree = renderer.create(<MeridiemPicker onSelect={onSelect} />);
    });
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('should select a meridiem when click', async () => {
    const { user } = renderDefault(<MeridiemPicker onSelect={onSelect} />);

    const pm = screen.getByText(testTranslate('MERIDIEM_PM', 'common'));
    expect(pm).toBeInTheDocument();
    expect(pm).toBeVisible();

    await user.click(pm);

    expect(onSelect).toHaveBeenCalledTimes(1);
    expect(onSelect).toHaveBeenCalledWith(false);
  });
});
