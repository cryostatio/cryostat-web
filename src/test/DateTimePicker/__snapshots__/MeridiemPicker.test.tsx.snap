// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MeridiemPicker/> renders correctly 1`] = `
<div
  aria-label="Select AM or PM"
  className="datetime-picker__meridiem-title-stack"
  role="listbox"
>
  <div
    className="datetime-picker__meridiem-tile selected"
    onClick={[Function]}
  >
    AM
  </div>
  <div
    className="datetime-picker__meridiem-tile unselected"
    onClick={[Function]}
  >
    PM
  </div>
</div>
`;
